Create method for import html

alphp opened the pull request:
loadHtml parse html string and create either SimpleXmlElement object or a DOMDocument.
use of empty($options['loadEntities']) instead of !$options['loadEntities']

alphp committed:
Create method for import html

loadHtml parse html string and create either SimpleXmlElement object or a DOMDocument.
use of empty($options['loadEntities']) instead of !$options['loadEntities']

stickler-ci committed:
Fixing style errors.

alphp committed:
Fixing style errors.

markstory commented on src/Utility/Xml.php:
```diff
@@ -162,7 +162,46 @@ protected static function _loadXml($input, $options)
+     * @return \SimpleXMLElement|\DOMDocument
+     * @throws \Cake\Utility\Exception\XmlException
+     */
+    public static function loadHtml($input, $options)
```
New methods require tests to be merged.

markstory commented:
Moved to 3.next as new features do not go into bugfix releases.

markstory commented on src/Utility/Xml.php:
```diff
@@ -142,7 +142,7 @@ protected static function _loadXml($input, $options)
         $hasDisable = function_exists('libxml_disable_entity_loader');
         $internalErrors = libxml_use_internal_errors(true);
-        if ($hasDisable && !$options['loadEntities']) {
+        if ($hasDisable && empty($options['loadEntities'])) {
```
>  use of empty($options['loadEntities']) instead of !$options['loadEntities']

Why though? What does this change?

alphp replied:
If not exists $options['loadEntities'] Notice is raised:
```
2018-06-02 02:48:02 Notice: Notice (8): Undefined index: loadEntities in [C:\usr\0xFF\bambu\app\vendor\cakephp\cakephp\src\Utility\Xml.php, line 184]
```

markstory replied:
How would that happen? The public methods set a default value.

alphp replied:
Ok, I see.
I based loadHtml in loadXml thats is calling by build, so in loadHtml $options don't have default value.

alphp committed:
Setting default values for $options in loadHtml()

Back to !$options['loadEntities'] instead of empty($options['loadEntities']).

alphp committed:
Add test for loadHtml method

alphp committed:
Fix sample.html

alphp committed:
Test parseHuge and Exception in loadHtml

alphp committed:
Fix testLoadHtml

markstory commented on tests/TestCase/Utility/XmlTest.php:
```diff
@@ -247,6 +247,39 @@ public function testBuildEmptyTag()
+        $xml = html_entity_decode($xml->asXML(), ENT_NOQUOTES, 'UTF-8');
+        $this->assertEquals($html, $xml);
+
+        $xml = Xml::loadHtml($html, ['parseHuge' => true]);
```
These tests are looking good. There should be a teat case covering the domdocument return tyoe as well.

alphp committed:
Update sample.html and testLoadHtml

Update sample.html to use HTML5 sample.html with doctype.
Update testLoadHtml to detect the presence of elements of sample.html and test if its content is correct.

stickler-ci committed:
Fixing style errors.

alphp committed:
testLoadHtml generate original html string from xml

Tests:
- Convert a html string to SimpleXmlElement
- Access to html nodes in SimpleXmlElement
- Nodes content
- Regenerate the original  html string from SimpleXmlElement

alphp committed:
testLoadHtml: Fix regenerate of original html string

alphp committed:
Revert "Fixing style errors."

This reverts commit 6572d67.

alphp committed:
testLoadHtml generate original html string from xml

Tests:
- Convert a html string to SimpleXmlElement
- Access to html nodes in SimpleXmlElement
- Nodes content
- Regenerate the original  html string from Xml Object
